### PR TITLE
[GLUTEN-3148][VL] Set minimum spill size for spillers to avoid paying zero bytes back to Java due to native reservation blocks

### LINKED ISSUE #3148: [VL] Unreasonable OOMs in isolated memory mode
### Backend

VL (Velox)

### Bug description

There are some nodes in consumer tree not getting spilled although they seem definitely spillable at the time of OOM.

Some logs collected:

```
23/09/14 05:30:52 INFO memory.TaskMemoryManager: Memory used in task 98549
23/09/14 05:30:52 INFO memory.TaskMemoryManager: Acquired by io.glutenproject.memory.memtarget.spark.TreeMemoryConsumer@9d68491: 2046.4 MiB
23/09/14 05:30:52 INFO memory.TaskMemoryManager: 0 bytes of memory were used by task 98549 but are not associated with specific consumers
23/09/14 05:30:52 INFO memory.TaskMemoryManager: 2208720485 bytes of memory are used for execution and 398257815 bytes of memory are used for storage
23/09/14 05:30:52 ERROR nmm.ManagedReservationListener: Error reserving memory from target
io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget$OutOfMemoryException: Not enough spark off-heap execution memory. Acquired: 8388608, granted: 1677722. Try tweaking config option spark.memory.offHeap.size to get larger space to run this application.
Current config settings:
        spark.gluten.memory.offHeap.size.in.bytes=16.0 GiB
        spark.gluten.memory.task.offHeap.size.in.bytes=2.0 GiB
        spark.gluten.memory.conservative.task.offHeap.size.in.bytes=2.0 GiB
Memory consumer stats:
        Task.98549:                                                       Current used bytes: 2046.4 MiB, peak bytes:        N/A
        \- Gluten.Tree.988:                                               Current used bytes: 2046.4 MiB, peak bytes:    2.0 GiB
           \- root.988, 2.0 GiB:                                          Current used bytes: 2046.4 MiB, peak bytes:    2.0 GiB
              +- WholeStageIterator.1003:                                 Current used bytes: 1568.0 MiB, peak bytes: 1576.0 MiB
              |  \- single:                                               Current used bytes: 1478.0 MiB, peak bytes:        N/A
              |     +- velox:                                             Current used bytes: 1478.0 MiB, peak bytes: 1564.0 MiB
              |     |  +- task.Gluten stage-465 task-98549:               Current used bytes: 1478.0 MiB, peak bytes: 1564.0 MiB
              |     |  |  +- node.3:                                      Current used bytes: 1440.0 MiB, peak bytes: 1480.0 MiB
              |     |  |  |  \- op.3.0.0.OrderBy:                         Current used bytes: 1333.0 MiB, peak bytes: 1479.8 MiB
              |     |  |  +- node.12:                                     Current used bytes:   20.0 MiB, peak bytes:   96.0 MiB
              |     |  |  |  \- op.12.0.0.PartialAggregation:             Current used bytes:   18.1 MiB, peak bytes:   89.5 MiB
              |     |  |  +- node.15:                                     Current used bytes:   13.0 MiB, peak bytes:   24.0 MiB
              |     |  |  |  \- op.15.0.0.Aggregation:                    Current used bytes:    3.1 MiB, peak bytes:   23.8 MiB
              |     |  |  +- node.8:                                      Current used bytes:    2.0 MiB, peak bytes:    2.0 MiB
              |     |  |  |  +- op.8.2.0.HashBuild:                       Current used bytes:   76.0 KiB, peak bytes:   76.0 KiB
              |     |  |  |  \- op.8.0.0.HashProbe:                       Current used bytes:   72.0 KiB, peak bytes:  168.0 KiB
              |     |  |  +- node.11:                                     Current used bytes: 1024.0 KiB, peak bytes: 1024.0 KiB
              |     |  |  |  \- op.11.0.0.FilterProject:                  Current used bytes:   12.3 KiB, peak bytes:   97.5 KiB
              |     |  |  +- node.17:                                     Current used bytes: 1024.0 KiB, peak bytes: 1024.0 KiB
              |     |  |  |  \- op.17.0.0.PartialAggregation:             Current used bytes:   64.3 KiB, peak bytes:   64.3 KiB
              |     |  |  +- node.14:                                     Current used bytes: 1024.0 KiB, peak bytes: 1024.0 KiB
              |     |  |  |  \- op.14.0.0.FilterProject:                  Current used bytes:    768.0 B, peak bytes:    768.0 B
              |     |  |  +- node.1:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.1.1.0.ValueStream:                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.10:                                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.10.0.0.FilterProject:                  Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.13:                                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.13.0.0.FilterProject:                  Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.6:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.6.0.0.FilterProject:                   Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.0:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.0.0.0.ValueStream:                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.7:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.7.0.0.FilterProject:                   Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.2:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.2.2.0.ValueStream:                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.4:                                      Current used bytes:      0.0 B, peak bytes:   20.0 MiB
              |     |  |  |  \- op.4.1.0.OrderBy:                         Current used bytes:      0.0 B, peak bytes:   10.1 MiB
              |     |  |  +- node.N/A:                                    Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.N/A.1.0.CallbackSink:                  Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.16:                                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.16.0.0.FilterProject:                  Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.5:                                      Current used bytes:      0.0 B, peak bytes: 1024.0 KiB
              |     |  |  |  \- op.5.0.0.MergeJoin:                       Current used bytes:      0.0 B, peak bytes:  146.3 KiB
              |     |  |  \- node.9:                                      Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |     \- op.9.0.0.FilterProject:                   Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  \- WholeStageIterator_default_leaf:                Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     \- arrow:                                             Current used bytes:      0.0 B, peak bytes:        N/A
              +- OverAcquire.DummyTarget.2448:                            Current used bytes:  470.4 MiB, peak bytes:  472.0 MiB
              +- ArrowContextInstance.625:                                Current used bytes:    8.0 MiB, peak bytes:    8.0 MiB
              +- OverAcquire.DummyTarget.2447:                            Current used bytes:      0.0 B, peak bytes:    2.4 MiB
              +- ShuffleWriter.1088:                                      Current used bytes:      0.0 B, peak bytes: 1638.4 KiB
              |  \- single:                                               Current used bytes:      0.0 B, peak bytes:        N/A
              |     +- arrow:                                             Current used bytes:      0.0 B, peak bytes:        N/A
              |     \- velox:                                             Current used bytes:      0.0 B, peak bytes:      0.0 B
              |        \- ShuffleWriter_default_leaf:                     Current used bytes:      0.0 B, peak bytes:      0.0 B
              +- OverAcquire.DummyTarget.3412:                            Current used bytes:      0.0 B, peak bytes:      0.0 B
              +- BuildSideRelation#BatchSerializer.579:                   Current used bytes:      0.0 B, peak bytes:    8.0 MiB
              |  \- single:                                               Current used bytes:      0.0 B, peak bytes:        N/A
              |     +- velox:                                             Current used bytes:      0.0 B, peak bytes: 1024.0 KiB
              |     |  \- BuildSideRelation#BatchSerializer_default_leaf: Current used bytes:      0.0 B, peak bytes:    6.0 KiB
              |     \- arrow:                                             Current used bytes:      0.0 B, peak bytes:        N/A
              +- OverAcquire.DummyTarget.2449:                            Current used bytes:      0.0 B, peak bytes:    2.4 MiB
              \- ShuffleReader.71:                                        Current used bytes:      0.0 B, peak bytes:    8.0 MiB
                 \- single:                                               Current used bytes:      0.0 B, peak bytes:        N/A
                    +- arrow:                                             Current used bytes:      0.0 B, peak bytes:        N/A
                    \- velox:                                             Current used bytes:      0.0 B, peak bytes:      0.0 B
                       \- ShuffleReader_default_leaf:                     Current used bytes:      0.0 B, peak bytes:      0.0 B
```

```
23/09/14 05:23:07 INFO memory.TaskMemoryManager: Memory used in task 88388
23/09/14 05:23:07 INFO memory.TaskMemoryManager: Acquired by io.glutenproject.memory.memtarget.spark.TreeMemoryConsumer@cfec2a5: 1578.4 MiB
23/09/14 05:23:07 INFO memory.TaskMemoryManager: 0 bytes of memory were used by task 88388 but are not associated with specific consumers
23/09/14 05:23:07 INFO memory.TaskMemoryManager: 14664125640 bytes of memory are used for execution and 167495645 bytes of memory are used for storage
23/09/14 05:23:07 ERROR nmm.ManagedReservationListener: Error reserving memory from target
io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget$OutOfMemoryException: Not enough spark off-heap execution memory. Acquired: 8388608, granted: 1677722. Try tweaking config option spark.memory.offHeap.size to get larger space to run this application.
Current config settings:
        spark.gluten.memory.offHeap.size.in.bytes=16.0 GiB
        spark.gluten.memory.task.offHeap.size.in.bytes=2.0 GiB
        spark.gluten.memory.conservative.task.offHeap.size.in.bytes=2.0 GiB
Memory consumer stats:
        Task.88388:                                         Current used bytes: 1578.4 MiB, peak bytes:        N/A
        \- Gluten.Tree.835:                                 Current used bytes: 1578.4 MiB, peak bytes:    2.0 GiB
           \- root.835, 2.0 GiB:                            Current used bytes: 1578.4 MiB, peak bytes:    2.0 GiB
              +- WholeStageIterator.842:                    Current used bytes: 1560.0 MiB, peak bytes: 1561.6 MiB
              |  \- single:                                 Current used bytes: 1176.0 MiB, peak bytes:        N/A
              |     +- velox:                               Current used bytes: 1176.0 MiB, peak bytes: 1560.0 MiB
              |     |  +- task.Gluten stage-322 task-88388: Current used bytes: 1176.0 MiB, peak bytes: 1560.0 MiB
              |     |  |  +- node.1:                        Current used bytes: 1176.0 MiB, peak bytes: 1176.0 MiB
              |     |  |  |  \- op.1.0.0.Aggregation:       Current used bytes: 1176.0 MiB, peak bytes: 1176.0 MiB
              |     |  |  +- node.0:                        Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  |  \- op.0.0.0.ValueStream:       Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |  +- node.3:                        Current used bytes:      0.0 B, peak bytes:  384.0 MiB
              |     |  |  |  \- op.3.0.0.OrderBy:           Current used bytes:      0.0 B, peak bytes:  317.7 MiB
              |     |  |  \- node.2:                        Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  |     \- op.2.0.0.FilterProject:     Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     |  \- WholeStageIterator_default_leaf:  Current used bytes:      0.0 B, peak bytes:      0.0 B
              |     \- arrow:                               Current used bytes:      0.0 B, peak bytes:        N/A
              +- ShuffleReader.45:                          Current used bytes:    8.0 MiB, peak bytes:    8.0 MiB
              |  \- single:                                 Current used bytes:   62.4 KiB, peak bytes:        N/A
              |     +- arrow:                               Current used bytes:   62.4 KiB, peak bytes:        N/A
              |     \- velox:                               Current used bytes:      0.0 B, peak bytes:      0.0 B
              |        \- ShuffleReader_default_leaf:       Current used bytes:      0.0 B, peak bytes:      0.0 B
              +- ArrowContextInstance.571:                  Current used bytes:    8.0 MiB, peak bytes:    8.0 MiB
              +- OverAcquire.DummyTarget.2080:              Current used bytes:    2.4 MiB, peak bytes:    2.4 MiB
              \- OverAcquire.DummyTarget.2079:              Current used bytes:      0.0 B, peak bytes:  468.0 MiB

        at io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget.borrow(ThrowOnOomMemoryTarget.java:88)
        at io.glutenproject.memory.nmm.ManagedReservationListener.reserve(ManagedReservationListener.java:42)

```

### Spark version

None

### Spark configurations

_No response_

### System information

_No response_

### Relevant logs

_No response_

### PR DESCRIPTION
**Pull request by zhztheplayer**
Minimum spill target size should be larger than spark.gluten.memory.reservationBlockSize，since any release action within size smaller than the block size may not have chance to report back to the Java-side reservation listener.

This would also fix #3148

Also rename two config options:

spark.gluten.sql.columnar.backend.velox.overAcquiredMemoryRatio -> spark.gluten.memory.overAcquiredMemoryRatio
spark.gluten.sql.columnar.backend.velox.reservationBlockSize -> spark.gluten.memory.reservationBlockSize

The options were not necessarily specific to Velox backend since they were in `gluten-data` module.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@Yohahaha I'll rename config option `spark.gluten.sql.columnar.backend.velox.overAcquiredMemoryRatio` to `spark.gluten.memory.overAcquiredMemoryRatio` in the path. Would you have any thoughts about that?

File: gluten-data/src/main/java/io/glutenproject/memory/nmm/NativeMemoryManagers.java [hunk 1]
```diff
@@ -62,28 +62,31 @@ public static NativeMemoryManager create(String name, Spiller spiller) {
...
+                                            new IllegalStateException(
+                                                ""
+                                                    + "Shrink is requested before native "
+                                                    + "memory manager is created. Try moving any "
+                                                    + "actions about memory allocation out "
+                                                    + "from the memory manager constructor.")),
+                            GlutenConfig.getConf().memoryReservationBlockSize()),
+                        // the input spiller, called after nmm.shrink was called
+                        Spillers.withMinSpillSize(
+                            spiller, GlutenConfig.getConf().memoryReservationBlockSize())),
```
Comment (Yohahaha) [inline gluten-data/src/main/java/io/glutenproject/memory/nmm/NativeMemoryManagers.java:89]: nit: declare and pass local variable for `GlutenConfig.getConf().memoryReservationBlockSize()`

**Comment by Yohahaha** [comment]
LGTM, thanks!

**Comment by Yohahaha** [comment]
Curious why remove `.sql.columnar.` parts?

**Comment by zhztheplayer** [comment]
> .sql.columnar.

Seems to me that we generally had two patterns of config naming:

`spark.gluten.*` and `spark.gluten.sql.columnar.*`

I think we just use the former one for the 2 configs, it's fine and not too tedious,

I even prefer eliminating all `.sql.columnar` from all existing config option names but that change will be extremely breaking.

File: gluten-core/src/main/java/io/glutenproject/memory/memtarget/spark/TreeMemoryConsumer.java [hunk 2]
```diff
@@ -220,23 +225,21 @@ public Spiller getNodeSpiller() {
...
+        if (spilledBytes == 0) {
           // OOM
           return false;
         }
-        prevFreeBytes = freeBytes;
-        // spill
-        long bytesToSpill = bytesNeeded - freeBytes;
-        spillTree(this, bytesToSpill);
       }
     }
```
Comment (zhztheplayer) [inline gluten-core/src/main/java/io/glutenproject/memory/memtarget/spark/TreeMemoryConsumer.java:244]: Note: This part of change can individually fix #3148 .

**Comment by Yohahaha** [comment]
> I even prefer eliminating all .sql.columnar from all existing config option names but that change will be extremely breaking

+1 on that, we may have a discussion in issue.

**Review by zhouyuan — APPROVED** [review]
👍

**Pull request merged**